substrate/in-instructions/primitives/src/lib.rs

```diff
@@ -0,0 +1,47 @@
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use scale::{Encode, Decode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+
+use sp_core::{ConstU32, bounded::BoundedVec};
+
+// Monero, our current longest address candidate, has a longest address of featured with payment ID
+// 1 (enum) + 1 (flags) + 64 (two keys) + 8 (payment ID) = 74
+pub const MAX_ADDRESS_LEN: u32 = 74;
+// Should be enough for a Uniswap v3 call
+pub const MAX_DATA_LEN: u32 = 512;
+
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct ExternalAddress(BoundedVec<u8, ConstU32<{ MAX_ADDRESS_LEN }>>);
+impl ExternalAddress {
+  #[cfg(feature = "std")]
+  pub fn new(address: Vec<u8>) -> Result<ExternalAddress, &'static str> {
+    Ok(ExternalAddress(address.try_into().map_err(|_| "address length exceeds {MAX_ADDRESS_LEN}")?))
```

Comment:
let's move to a validation function call for external addresses? We probably need to add a lot of checks.

Reply:
We can't validate EAs here right now. Validation, which is on purposely defined as non-existent at this step, requires context of the coin we're running against and requires a provided BTC/ETH/XMR address format. This is agnostic to that by just being the raw bytes.

The processor should perform validation when converting from Vec<u8> to the actual type IMO.

Reply:
since validation is simple here we should to it every step but wont fight you on it.

Reply:
> requires context of the coin we're running against and requires a provided BTC/ETH/XMR address format.

It's impossible here unless every single ExternalAddress is ExternalAddress<C: Coin>, which then forces every single Instruction to be C: Coin, which breaks how it's intended to be a single type usable for all integrations since it's now type-locked to a specific coin. We do not have the context and cannot safely provide the context at this level.